polkadot/network/src/consensus.rs

```diff
@@ -0,0 +1,301 @@
+	Ok(Some(match msg.message {
+		msg::BftMessage::Consensus(c) => bft::generic::Communication::Consensus(match c {
+			msg::SignedConsensusMessage::Propose(proposal) => bft::generic::LocalizedMessage::Propose({
+				if &proposal.sender == local_id { return Ok(None) }
+				let proposal = bft::generic::LocalizedProposal {
+					round_number: proposal.round_number as usize,
+					proposal: proposal.proposal,
+					digest: proposal.digest,
+					sender: proposal.sender,
+					digest_signature: ed25519::LocalizedSignature {
+						signature: proposal.digest_signature,
+						signer: ed25519::Public(proposal.sender.into()),
+					},
+					full_signature: ed25519::LocalizedSignature {
+						signature: proposal.full_signature,
+						signer: ed25519::Public(proposal.sender.into()),
+					}
+				};
+				bft::check_proposal(authorities, &msg.parent_hash, &proposal)?;
+
+				trace!(target: "bft", "importing proposal message for round {} from {}", proposal.round_number, Hash::from(proposal.sender.0));
+				proposal
+			}),
+			msg::SignedConsensusMessage::Vote(vote) => bft::generic::LocalizedMessage::Vote({
+				if &vote.sender == local_id { return Ok(None) }
+				let vote = bft::generic::LocalizedVote {
+					sender: vote.sender,
+					signature: ed25519::LocalizedSignature {
+						signature: vote.signature,
+						signer: ed25519::Public(vote.sender.0),
+					},
+					vote: match vote.vote {
+						msg::ConsensusVote::Prepare(r, h) => bft::generic::Vote::Prepare(r as usize, h),
+						msg::ConsensusVote::Commit(r, h) => bft::generic::Vote::Commit(r as usize, h),
+						msg::ConsensusVote::AdvanceRound(r) => bft::generic::Vote::AdvanceRound(r as usize),
+					}
+				};
+				bft::check_vote::<Block>(authorities, &msg.parent_hash, &vote)?;
+
+				trace!(target: "bft", "importing vote {:?} from {}", vote.vote, Hash::from(vote.sender.0));
+				vote
+			}),
+		}),
+		msg::BftMessage::Auxiliary(a) => {
+			let justification = bft::UncheckedJustification::from(a);
+			// TODO: get proper error
+			let justification: Result<_, bft::Error> = bft::check_prepare_justification::<Block>(authorities, msg.parent_hash, justification)
+				.map_err(|_| bft::ErrorKind::InvalidJustification.into());
+			bft::generic::Communication::Auxiliary(justification?)
+		},
+	}))
+}
+
+// task that processes all gossipped consensus messages,
+// checking signatures
+struct MessageProcessTask<P: PolkadotApi> {
+	inner_stream: mpsc::UnboundedReceiver<ConsensusMessage<Block>>,
+	bft_messages: mpsc::UnboundedSender<bft::Communication<Block>>,
+	validators: Vec<SessionKey>,
+	table_router: Router<P>,
+}
+
+impl<P: LocalPolkadotApi + Send + Sync + 'static> Future for MessageProcessTask<P> {
+	type Item = ();
+	type Error = ();
+
+	fn poll(&mut self) -> Poll<(), ()> {
+		loop {
+			match self.inner_stream.poll() {
+				Ok(Async::Ready(Some(val))) => match val {
```

Comment:
maybe introduce a `process_message` auxiliary to avoid quite such an indentation depth?